# Homework 7

## Your Name Here (or your names here if you are pair programming)

Student Name: Tyler Smedley

Student UT EID: tws933

---

Partner Name:

Partner UT EID:

---

Date Created: 03/18/24

Date Last Modified:

---

Totoal Points 20. 



## Supprt Vector Machine 

In [2]:
# Standard Headers
# You are welcome to add additional headers here if you wish
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

# Enable inline mode for matplotlib so that Jupyter displays graphs
%matplotlib inline

# Your allowed to use only the above libraries that are imported. No other libs should be used in this assignment. 

## Heart Dataset 

In this Assignment we will work with some patients dataset. 

We have access to 303 patients data. The features are listed below. 

In [3]:
# Your code here
heart_df = pd.read_csv("Heart.csv")
heart_df

Age  Sex     ChestPain  RestBP  Chol  Fbs  RestECG  MaxHR  ExAng  \
0     63    1       typical     145   233    1        2    150      0   
1     67    1  asymptomatic     160   286    0        2    108      1   
2     67    1  asymptomatic     120   229    0        2    129      1   
3     37    1    nonanginal     130   250    0        0    187      0   
4     41    0    nontypical     130   204    0        2    172      0   
..   ...  ...           ...     ...   ...  ...      ...    ...    ...   
298   45    1       typical     110   264    0        0    132      0   
299   68    1  asymptomatic     144   193    1        0    141      0   
300   57    1  asymptomatic     130   131    0        0    115      1   
301   57    0    nontypical     130   236    0        2    174      0   
302   38    1    nonanginal     138   175    0        0    173      0   

     Oldpeak  Slope   Ca        Thal Target  
0        2.3      3  0.0       fixed     No  
1        1.5      2  3.0      normal    Yes  
2        2.6      2  2.0  reversable    Yes  
3        3.5      3  0.0      normal     No  
4        1.4      1  0.0      normal     No  
..       ...    ...  ...         ...    ...  
298      1.2      2  0.0  reversable    Yes  
299      3.4      2  2.0  reversable    Yes  
300      1.2      2  1.0  reversable    Yes  
301      0.0      2  1.0      normal    Yes  
302      0.0      1  NaN      normal     No  

[303 rows x 14 columns]

**Age:** The person’s age in years

**Sex:** The person’s sex (1 = male, 0 = female)

**ChestPain:** chest pain type

* Value 0: asymptomatic
* Value 1: atypical angina
* Value 2: non-anginal pain
* Value 3: typical angina

**RestBP:** The person’s resting blood pressure (mm Hg on admission to the hospital)

**Chol:** The person’s cholesterol measurement in mg/dl

**Fbs:** The person’s fasting blood sugar (> 120 mg/dl, 1 = true; 0 = false)
restecg: resting electrocardiographic results

* Value 0: showing probable or definite left ventricular hypertrophy by Estes’ criteria
* Value 1: normal
* Value 2: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV)

**RestECG:** The person’s maximum heart rate achieved

**MaxHR:** Exercise induced angina (1 = yes; 0 = no)

**Oldpeak:** ST depression induced by exercise relative to rest (‘ST’ relates to positions on the ECG plot. See more here)

**Slope:** the slope of the peak exercise ST segment — 0: downsloping; 1: flat; 2: upsloping

* 0: downsloping; 
* 1: flat; 
* 2: upsloping

**Ca:** The number of major vessels (0–3)

**Thal:** A blood disorder called thalassemia Value 0: NULL (dropped from the dataset previously

* Value 1: fixed defect (no blood flow in some part of the heart)
* Value 2: normal blood flow
* Value 3: reversible defect (a blood flow is observed but it is not normal)

**Target:** Heart disease (1 = no, 0= yes)

# Task - 1 Implement SVM using libraries (4 points)
We want to use **Suppert Vector Machine** to perdict if the patients will have heart problems or not. The column "Target" in our datasets includes data about heart diseases. If the patient had heart disease we have a 1 and if not a zero. 

Prepare your data set for predicting heart disease ("Target" column) out of 3 features:

* Age of the patient (Column **"Age"**)
* Gender of the patient (male or female - Column **"Sex"**)
* Cholestrol level of the patient (Column **"Chol"**) 


Split your data into 80% traning data and 20% test data, and implement Support Vector Machine using Scikit-Learn. 





In [4]:
# Convert target yes/no to 1/0 and create test/train
from sklearn.model_selection import train_test_split

heart_df['Target'] = heart_df['Target'].map({'Yes':1, 'No':0})
y = heart_df['Target']
X = heart_df[['Age', 'Sex', 'Chol']]

y_train, y_test, X_train, X_test = train_test_split(y, X, test_size=0.2, random_state=0)

In [5]:
# import sklearn's svm model, fit training data, and predict target values from test data
from sklearn import svm

svm_model  = svm.SVC()
svm_model.fit(X_train, y_train)

target_pred = svm_model.predict(X_test)

# Task 2 - (4 points)

Cacluate the accuracy, Precision, Recall and F1 score of your **SVM** implementaion from Task 1. 
Print the results. 

You may use library methods for this task if you choose to.


In [6]:
# use sklearn to calculate and print accuracy, precision, recall, and F1 score
from sklearn.metrics import classification_report

print(classification_report(y_test, target_pred))

              precision    recall  f1-score   support

           0       0.56      0.69      0.62        35
           1       0.39      0.27      0.32        26

    accuracy                           0.51        61
   macro avg       0.47      0.48      0.47        61
weighted avg       0.49      0.51      0.49        61



# Task 3 - Implement SVM without using libraries  - (4 points)

Implement SVM from scratch using Hinge Loss function and Gradient Descent. 
Try to produce the same result as you get from the libraries. 


* Do as many iterations as needed 
* Do maximum **100 iterations**
* Use a very small learning rate for checking your GD implementation. 
* Your are allowed to use your choice of learning rate, like using 0.0001, 0.001 or 0.01 or 0.1 or higher. 
* Visualize your costs. 
* No need to add an y-intercept in this task. 
* You can use libraries to report accuracy, Precision, Recall and F1. 



In [7]:
#TODO:
    # manualy implement train_test_split
    # convert categorical variables to numerical
    # Convert X data frame to numpy style array

In [13]:
import random
# creating a new data frame for the manual implementation
heart_2 = pd.read_csv('Heart.csv')
heart_2['Target'] = heart_2['Target'].map({'Yes':1, 'No': -1})
random_test_indices = random.sample(range(0, heart_2.shape[0]), int(heart_2.shape[0] * 0.2))


[156,
 49,
 46,
 133,
 129,
 290,
 194,
 199,
 246,
 301,
 250,
 127,
 11,
 242,
 182,
 39,
 21,
 230,
 105,
 151,
 179,
 14,
 161,
 104,
 296,
 283,
 73,
 3,
 140,
 269,
 13,
 63,
 202,
 146,
 95,
 214,
 286,
 25,
 38,
 86,
 291,
 60,
 229,
 292,
 260,
 113,
 239,
 71,
 31,
 20,
 82,
 267,
 37,
 10,
 125,
 274,
 191,
 256,
 219,
 244]

In [9]:
# function to compute cost here
def compute_cost(X, y, W, reg_factor):
    n = X.shape[1]
    distances = 1 - y *(np.dot(X, W))
    distances[distances < 0] = 0

    hinge_loss = reg_factor * (np.sum(distances) / n)
    return (np.dot(W, W) + hinge_loss)


In [10]:
# function to calculate gradient
def calculate_gradient(X, y, W, reg_factor):
    if type(y) == np.float64:
        y = np.array([y])
        X = np.array([X])

    distance = 1 - (y * np.dot(X, W))
    dw = np.zeros(len(W))

    for ind, d  in enumerate(distance):
        if d < 0:
            di = W
        else:
            di = W - (reg_factor * y[ind] * X[ind])
        dw += di

    dw = dw/len(y)
    return dw



In [11]:
weights  = np.zeros(X.shape[1])
y_train = np.where(y_train==0, -1, 1)
print(X_train)

# optimize SVM using gradient descent
num_iterations = 100
learning_rate = 0.01
regularization = 0.01
cost_list = []

for i in range(0, num_iterations):
    cost = compute_cost(X_train, y_train, weights, regularization)
    print("Epoch", i, "Cost is:", cost, "weights", weights)
    cost_list.append(cost)
    gradient = calculate_gradient(X_train, y_train, weights, regularization)
    
    weights = weights - learning_rate * gradient


     Age  Sex  Chol
74    44    1   197
153   55    1   289
64    54    1   188
296   59    1   176
287   58    1   220
..   ...  ...   ...
251   58    1   218
192   43    1   247
117   35    0   183
47    50    1   243
172   59    0   249

[242 rows x 3 columns]
Epoch 0 Cost is: 0.8066666666666668 weights [0. 0. 0.]


KeyError: 0

# Task 4 - Compare SVM results with Logistic Regression - (4 points)

Which model performs better here? Compare your results wit the logistic regression. You can use libraries for this task, it is not necessary to implement logistic regression from sratch.


In [ ]:
from sklearn.linear_model import LogisticRegression
lr_model = LogisticRegression()

lr_model.fit(X_train, y_train)
lr_predict = lr_model.predict(X_test)

print(classification_report(y_test, lr_predict))
# The logistic regression model scored better in every metric than the svm model
# the accuarcy of svm was about 50% while the logistic regression was about 75%

              precision    recall  f1-score   support

          -1       0.00      0.00      0.00         0
           0       0.00      0.00      0.00        35
           1       0.69      0.77      0.73        26

    accuracy                           0.33        61
   macro avg       0.23      0.26      0.24        61
weighted avg       0.29      0.33      0.31        61



/home/tylers/anaconda3/envs/env_full/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/tylers/anaconda3/envs/env_full/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Recall and F-score are ill-defined and being set to 0.0 in labels with no true samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/home/tylers/anaconda3/envs/env_full/lib/python3.11/site-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(res

# Task 5 - Apply a kernel function to improve SVM performance (4 points)

Use the Scikit-learn librariy and apply a kernel function to improve the SVM performance. Check if this is possible. 


In [ ]:
# Add your code Here! 